### PR TITLE
Resolve variable names at parsing time

### PR DESCRIPTION
**Pull request by jeremiedimino**
Currently, at parsing time variables of the form `%{project_root}` are kept as strings along side with the version of the dune language in use. When comes the time to evaluate the variable, two steps happen:

1. the variable is resolved and converted to a `Pform.t` value, which is a symbolic representation of the variable. The resolution takes into account the variable name as well as the syntax version
2. the `Pform.t` value is then evaluated as appropriate

This means that when manipulating string with variables (`String_wiht_vars.t`), we must remember that the meaning of a variable name depends on the syntax version.

This PR attempts to move step 1 at parsing time, More precisely, right after parsing string with variables are now fully resolved and embed symbolic variable names that are independent of the language version rather than string that are dependant on the language version.

### CODE REVIEW

File: src/dune_lang/template.mli [hunk 1]
```diff
@@ -29,3 +29,26 @@ val pp_split_strings : Format.formatter -> t -> unit
 val remove_locs : t -> t
 
 val to_dyn : t -> Dyn.t
+
+module Var : sig
```
Comment (rgrinberg) [inline src/dune_lang/template.mli:33]: If you move this signature to the top, you can get rid of the `type t = var` alias.

File: src/dune_engine/pform.mli [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  | Var of Var.t
+  | Macro of Macro.t * string
+
+val compare : t -> t -> Ordering.t
+
+val to_dyn : t -> Dyn.t
+
+(** Enode using the latest name for a given percent form. Return [None] if the
+    percent form was deleted. *)
+val encode_to_latest_dune_lang_version : t -> (string * string option) option
```
Comment (rgrinberg) [inline src/dune_engine/pform.mli:97]: IMO this would be cleaner if there was a dedicated variant for the return value.

**Comment by rgrinberg** [comment]
Looks good. Its's a bit of a shame that more stuff is added to `Dune_engine`. I wonder if it makes sense to move `String_with_vars` and `Pform` all the way to `Dune_lang`.

File: src/dune_engine/pform.mli (hunk 2 above)
Comment (jeremiedimino) [reply src/dune_engine/pform.mli:103]: Agreed, changing that

**Comment by jeremiedimino** [comment]
Agreed, it's a bit unfortunate. We need to move `Cm_kind` and `Mode` as well, and moving them to `Dune_lang` doesn't feel right. We should probably introduce a specific library for all this stuff.

File: src/dune_lang/template.mli (hunk 1 above)
Comment (jeremiedimino) [reply src/dune_lang/template.mli]: Well spotted. I moved it and also moved a few pform related functions to it.

**Comment by jeremiedimino** [comment]
Alright, it should be ready now. One difference with before is that we can now detect unknown variables immediately, which seems better. To be conservative, I enabled this behaviour only for lang dune >= 3.0.

**Comment by rgrinberg** [comment]
Looks like this is still too eager:

```
#=== ERROR while compiling core_kernel.v0.14.1 ================================#
# context              2.0.7 | macos/x86_64 | ocaml-base-compiler.4.10.0 | git+https://github.com/ocaml/opam-repository.git
# path                 ~/.opam/4.10.0/.opam-switch/build/core_kernel.v0.14.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p core_kernel -j 5
# exit-code            1
# env-file             ~/.opam/log/core_kernel-42054-4f494f.env
# output-file          ~/.opam/log/core_kernel-42054-4f494f.out
### output ###
# File "version_util/test/bin/dune", line 8, characters 18-23:
# 8 |  (action (bash "%{NODE} main.bc.js > /dev/null")))
#                       ^^^^^
# Error: Unknown variable %{NODE}
```

Although this certainly looks like a bug. How could `%{NODE}` possibly work?

**Comment by jeremiedimino** [comment]
> Although this certainly looks like a bug in core_kernel. How could %{NODE} possibly work?

It doesn't. But tests in jane street packages are currently not tested. I thought I made it so this wasn't a hard error, let me check that.

**Comment by jeremiedimino** [comment]
Should be fixed now. I was raising too early in `partial_expand`

**Comment by rgrinberg** [comment]
This needs a change log update for 3.0

**Review by rgrinberg — APPROVED** [review]
The new code looks much better. I didn't review for correctness, because it would take too long and the tests are probably enough to demonstrate this is all correct.

As mentioned earlier, it's better to keep CHANGS updated now rather than fill it in later and invariably forget something.

**Comment by jeremiedimino** [comment]
Indeed, I updated the changelog

**Pull request merged**